### PR TITLE
Add table importing feature

### PR DESCRIPTION
**Pull request by jnmt**
This PR adds a table-importing feature to enable users to import their existing databases to ScalarDB. Sorry for the big PR.

You can import your existing databases using schema-loader with the `--import` option. For the `--import` option, you need to specify a schema JSON file with `-f`. The grammar of the schema file is basically the same as the usual loading, but it only requires a namespace, table, and transactional flag. Other items are ignored, and you don't have to specify columns since the import function automatically reads table metadata and recognizes columns.

Note that we do not support SQLite at this moment.  Rough policies of supported data types are as follows.

- We do not support data types that are larger or have higher precision than any of the ScalarDB data types.
  - The only exception is BIGINT. Our BIGINT is a little smaller than a regular 8-byte BIGINT due to the DynamoDB support, but I left BIGINT supported with a warning because we plan to remove this limitation (w/o DynamoDB).
- We notify users when we use a ScalarDB data type for an underlying data type that is smaller or has lower precision than ScalarDB's one.
- NUMERIC and DECIMAL are currently not supported.
- See inline comments for all supported (mapped data types) and unsupported data types, respectively.

### CODE REVIEW

File: core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java
Comment (jnmt) [inline core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java:236]: Refer to the original data types above and how they are expected to be mapped to ScalarDB data types. We probably need a document for supported and unsupported data types.

File: schema-loader/src/test/java/com/scalar/db/schemaloader/ImportSchemaParserTest.java
Comment (jnmt) [inline schema-loader/src/test/java/com/scalar/db/schemaloader/ImportSchemaParserTest.java:21]: This is an example of a schema JSON file for the table import.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitAdmin.java
Comment (jnmt) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitAdmin.java:197]: This is an entry point for a major change in this PR. It first gets JDBC metadata and creates table metadata for ScalarDB in `getImportTableMetadata()`. Then, it adds transaction metadata columns. Finally, it constructs correct ScalarDB metadata using the repair function.

File: core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java:473]: Just a quick question. Clustering keys aren't automatically created in this import feature, right?

File: core/src/main/java/com/scalar/db/util/ScalarDbUtils.java
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/util/ScalarDbUtils.java:257]: 100 and 101 are BINARY_FLOAT and BINARY_DOUBLE, respectively? How about adding comments about that?

File: core/src/main/java/com/scalar/db/util/ScalarDbUtils.java
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/util/ScalarDbUtils.java:254]: > int sqlTypes

`sqlType` is better?

File: schema-loader/src/main/java/com/scalar/db/schemaloader/ImportSchemaParser.java
Comment (komamitsu) [inline schema-loader/src/main/java/com/scalar/db/schemaloader/ImportSchemaParser.java:37]: [trivial] We don't need to use the concrete class type `ArrayList` as only `List#add` is used here.

```
    List<ImportTableSchema> tableSchemaList = new ArrayList<>();
```

File: schema-loader/src/main/java/com/scalar/db/schemaloader/ImportTableSchema.java
Comment (komamitsu) [inline schema-loader/src/main/java/com/scalar/db/schemaloader/ImportTableSchema.java:18]: [minor] How about adding `tableFullName` in this error message?

File: schema-loader/src/main/java/com/scalar/db/schemaloader/ImportTableSchema.java
Comment (komamitsu) [inline schema-loader/src/main/java/com/scalar/db/schemaloader/ImportTableSchema.java:16]: Is it expected if the length of `fullName` is 3 or more?

File: core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineOracle.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineOracle.java:249]: Oops, Oracle's BINARY_DOUBLE is compatible with ScalarDB's double (w.r.t. IEEE 754). So, we don't need to warn in the first place. Thank you!

File: core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java:473]: No, they aren't. Because the current table-importing feature targets relational databases only, clustering keys no longer be needed under the relational abstraction.

File: core/src/main/java/com/scalar/db/util/ScalarDbUtils.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/util/ScalarDbUtils.java:254]: Ahh, sqlTypes comes from `java.sql.Types`. Still weird? How about adding usage comments?

File: schema-loader/src/main/java/com/scalar/db/schemaloader/ImportTableSchema.java
Comment (jnmt) [reply schema-loader/src/main/java/com/scalar/db/schemaloader/ImportTableSchema.java:16]: Thanks. Should it be `length!= 2` right?

File: core/src/main/java/com/scalar/db/util/ScalarDbUtils.java
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/util/ScalarDbUtils.java:254]: Yeah, just looked at `java.sql.Types`. It consists of many `int` constant values (e.g., `public final static int SMALLINT =   5`.) So it's a set of integers (SQL types) in a sense.

`com.scalar.db.util.ScalarDbUtils#getJdbcType` seems to find the corresponding type out of the set of SQL types (`java.sql.Types`) by a single key `int sqlType(s)`. So I think we don't need to treat `int sqlType(s)` as plural.

File: core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java
Comment (komamitsu) [inline core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java:38]: `serial` should be added as well?

File: core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java
Comment (komamitsu) [inline core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java:54]: `float4` should be added for https://github.com/scalar-labs/scalardb/pull/931/files#diff-33515a82af5b0cb8f8944cc2237473e9eb8df95bca427831ddfdb72212184009R240 ?

File: core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java
Comment (komamitsu) [inline core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java:74]: This is for https://github.com/scalar-labs/scalardb/pull/931/files#diff-de99a8d73f3719cd7ae87026fbfbf7a953774fa4719bfdb18f898b5925e7d484R236? So how about testing with `FLOAT(52)` to test another validation?

Comment (komamitsu) [reply core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java:74]: Oops. The validation is just for logging. Please ignore it...

File: core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineOracle.java
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineOracle.java:274]: This should be `warn`?

Comment (jnmt) [reply core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineOracle.java:274]: I guess this should be `info`, but the message was wrong. Anyway, good catch! Thank you!

File: core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java
Comment (jnmt) [reply core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java:95]: Do you mean why we do not have to add the timestamp data type here? If yes, that is because there is no such data type in the reference. I'm not really sure, but I don't think `rowversion` is an alternative for `timestamp` since `datetime` and `datetime2` are mapped to JDBC `TIMESTAMP`. Note that the document you pointed out also says that:

> The Transact-SQL timestamp data type is different from the timestamp data type defined in the ISO standard.

Anyway, we don't have to have `timestamp` data type here, I think.

File: core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java
Comment (jnmt) [reply core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java:54]: No, we don't have to add it because the JDBC data type for PostrgreSQL `real` (`float4`) is always `REAL`.

File: core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java
Comment (jnmt) [reply core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java:38]: Oh, good catch! Thank you!

File: core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java
Comment (komamitsu) [reply core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java:95]: I see. Thanks for the explanation!

File: core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineOracle.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineOracle.java:274]: Let me take that back. I think I used the "smaller" message since the maximum size is larger than the logical maximum size of ScalarDB BLOB. So, this should be `warn` as you noticed.

https://docs.oracle.com/en/database/oracle/oracle-database/21/sqlrf/Data-Types.html#GUID-7B72E154-677A-4342-A1EA-C74C1EA928E6

> A binary large object. Maximum size is (4 gigabytes - 1) * (database block size).

File: core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java:452]: `RdbEngineStrategy` doesn't override `toString()`, so maybe `rdbEngine.getClass().getName()` is more readable:
```suggestion
          "Importing table is not allowed in this storage: " + rdbEngine.getClass().getName());
```

**Review by brfrn169 — COMMENTED** [review]
I've completed the first review and left a few comments. Please take a look when you have time! And I'll conduct the second review later. Thanks.

File: core/src/main/java/com/scalar/db/api/Admin.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/api/Admin.java:415]: That's a good question. Behind this question, are you wondering if we should only support transaction API for importing? It's a decision for what use cases should be covered. I originally thought covering both use cases (storage and transaction) would be nice. But, now I'm not really sure since the major reason for migrating to ScalarDB would be that users want to use transactions in ScalarDB. I would like to also hear other reviewers' thoughts.

File: core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineOracle.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineOracle.java]: Fixed in 0d3a2f8.

File: core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineSqlServer.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineSqlServer.java]: Fixed in 0d3a2f8.

File: core/src/main/java/com/scalar/db/util/ScalarDbUtils.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/util/ScalarDbUtils.java]: Fixed in 0d3a2f8.

File: core/src/main/java/com/scalar/db/util/ScalarDbUtils.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/util/ScalarDbUtils.java]: Fixed in 0d3a2f8.

File: schema-loader/src/main/java/com/scalar/db/schemaloader/ImportSchemaParser.java
Comment (jnmt) [reply schema-loader/src/main/java/com/scalar/db/schemaloader/ImportSchemaParser.java]: Fixed in 0d3a2f8.

File: schema-loader/src/main/java/com/scalar/db/schemaloader/ImportTableSchema.java
Comment (jnmt) [reply schema-loader/src/main/java/com/scalar/db/schemaloader/ImportTableSchema.java]: Fixed in 0d3a2f8.

File: schema-loader/src/main/java/com/scalar/db/schemaloader/ImportTableSchema.java
Comment (jnmt) [reply schema-loader/src/main/java/com/scalar/db/schemaloader/ImportTableSchema.java]: Fixed in 0d3a2f8.

File: core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java
Comment (jnmt) [reply core/src/integration-test/java/com/scalar/db/storage/jdbc/JdbcAdminImportTestUtils.java:38]: Fixed in 0d3a2f8.

File: core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineOracle.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/storage/jdbc/RdbEngineOracle.java]: Fixed in 0d3a2f8.

File: core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/storage/jdbc/JdbcAdmin.java]: Fixed in 0d3a2f8.

**Review by komamitsu — APPROVED** [review]
LGTM! 👍

**Review by brfrn169 — COMMENTED** [review]
I did the second review and I left some additional comments. Please take a look when you have time!

File: schema-loader/src/main/java/com/scalar/db/schemaloader/command/SchemaLoaderCommand.java
Comment (jnmt) [reply schema-loader/src/main/java/com/scalar/db/schemaloader/command/SchemaLoaderCommand.java]: Good catch, thank you! I totally forgot to fill it out. Fixed in 1526725.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitUtils.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitUtils.java]: Fixed in 1526725.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitUtils.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitUtils.java]: Fixed in 1526725.

File: core/src/main/java/com/scalar/db/transaction/jdbc/JdbcTransactionAdmin.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/transaction/jdbc/JdbcTransactionAdmin.java]: Exactly, we should. Fixed in 1526725.

File: core/src/main/java/com/scalar/db/util/ScalarDbUtils.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/util/ScalarDbUtils.java]: Yeah, it's also a good point. Thank you. Fixed in 1526725.

File: core/src/main/java/com/scalar/db/api/Admin.java
Comment (jnmt) [reply core/src/main/java/com/scalar/db/api/Admin.java:415]: @feeblefakie We are discussing whether we should support table importing in both storage and transaction API. What do you think?

**Review by Torch3333 — APPROVED** [review]
LGTM, thank you!

File: core/src/main/java/com/scalar/db/api/Admin.java
Comment (feeblefakie) [reply core/src/main/java/com/scalar/db/api/Admin.java:415]: @jnmt I think supporting only transactions seems fine to me.
Can we add support for storage if we really want to do it later on?

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!

File: core/src/main/java/com/scalar/db/api/Admin.java
Comment (feeblefakie) [reply core/src/main/java/com/scalar/db/api/Admin.java:415]: @jnmt I discuss it with Toshi, and we think we can keep it as is.
Later, we might remove the support for storage, though.

Comment (jnmt) [reply core/src/main/java/com/scalar/db/api/Admin.java:415]: @feeblefakie @Torch3333 Got it! Let's keep it as is. Thank you! (I was just wondering about the revert cost of the current PR.)

**Review by brfrn169 — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**
